chat-plugins/mafia.js

```diff
@@ -852,6 +915,194 @@ class MafiaTracker extends Rooms.RoomGame {
+
+		this.sendRoom(`IDEA roles have been distributed. You will have ${IDEA_TIMER / 1000} seconds to make your picks.`, {declare: true});
+		this.IDEA.timer = setTimeout(() => { this.ideaFinalizePicks(); }, IDEA_TIMER);
+
+		return ``;
+	}
+
+	/**
+	 *
+	 * @param {User} user
+	 * @param {string[]} selection
+	 */
+	ideaPick(user, selection) {
+		let buf = '';
+		if (this.phase !== 'IDEApicking') return 'The game is not in the IDEA picking phase.';
+		if (!this.IDEA || !this.IDEA.data) return this.sendRoom(`Trying to pick an IDEA role with no module running, target: ${JSON.stringify(selection)}. Please report this to a mod.`);
+		const player = this.players[user.userid];
+		if (!player.IDEA) return this.sendRoom(`Trying to pick an IDEA role with no player IDEA object, user: ${user.userid}. Please report this to a mod.`);
+		selection = selection.map(item => toId(item));
+		if (selection.length === 1 && this.IDEA.data.picks.length === 1) selection = [this.IDEA.data.picks[0], selection[0]];
+		if (selection.length !== 2) return user.sendTo(this.room, `|error|Invalid selection.`);
+
+		// input is formatted as ['selection', 'role']
+		// eg: ['role', 'bloodhound']
+		// ['alignment', 'alien']
+		// ['selection', ''] deselects
+		if (selection[1]) {
+			const roleIndex = player.IDEA.choices.map(choice => toId(choice)).indexOf(selection[1]);
+			if (roleIndex === -1) return user.sendTo(this.room, `|error|${selection[1]} is not an available role, perhaps it is already selected?`);
+			selection[1] = player.IDEA.choices.splice(roleIndex, 1)[0];
+		} else {
+			selection[1] = '';
+		}
+
+		if (player.IDEA.picks[selection[0]]) {
+			buf += `You have deselected ${player.IDEA.picks[selection[0]]}. `;
+			player.IDEA.choices.push(player.IDEA.picks[selection[0]]);
+		}
+
+		if (player.IDEA.picks[selection[0]] && !selection[1]) {
+			this.IDEA.waitingPick.push(player.userid);
+		} else if (!player.IDEA.picks[selection[0]] && selection[1]) {
+			this.IDEA.waitingPick.splice(this.IDEA.waitingPick.indexOf(player.userid), 1);
+		}
+
+		player.IDEA.picks[selection[0]] = selection[1];
+		if (selection[1]) buf += `You have selected ${selection[0]}: ${selection[1]}.`;
+		player.updateHtmlRoom();
+		if (!this.IDEA.waitingPick.length) {
+			if (this.IDEA.timer) clearTimeout(this.IDEA.timer);
+			this.ideaFinalizePicks();
+			return;
+		}
+		return user.sendTo(this.room, buf);
+	}
+
+	ideaFinalizePicks() {
+		if (!this.IDEA || !this.IDEA.data) return this.sendRoom(`Tried to finalize IDEA picks with no IDEA module running, please report this to a mod.`);
+		let randed = [];
+		for (const p in this.players) {
+			const player = this.players[p];
+			if (!player.IDEA) return this.sendRoom(`Trying to pick an IDEA role with no player IDEA object, user: ${player.userid}. Please report this to a mod.`);
+			let randPicked = false;
+			let role = [];
+			for (const choice of this.IDEA.data.picks) {
+				if (!player.IDEA.picks[choice]) {
+					randPicked = true;
```

Comment:
Why make a variable for this? just push them to the array of randomly picked players here.

Reply:
don't want them to be on the list multiple times if they didnt pick multiple things